fix(core): guard usages of `performance.mark`

alan-agius4 opened the pull request:
While `performance.mark` is available on all supported browsers and node.js version this API is not available in JSDOM which is used by Jest and Cloudflare worker.

This commit, updates the usage to a safer variant.

alan-agius4 committed:
fix(core): guard usages of `performance.mark`

While `performance.mark` is available on all supported browsers and node.js version this API is not available in JSDOM which is used by Jest and Cloudflare worker.

This commit, updates the usage to a safer variant.

devversion commented on packages/core/src/util/performance.ts:
```diff
@@ -0,0 +1,17 @@
+ */
+
+// tslint:disable:no-performance
+export function performanceMark(
```
Nit: I'd propose adding a method description, also explaining the reasoning on why this exists

devversion commented on tools/tslint/noPerformanceRule.ts:
```diff
@@ -0,0 +1,45 @@
+import {isIdentifier, isPropertyAccessExpression} from 'tsutils';
+import ts from 'typescript';
+
+export class Rule extends AbstractRule {
```
Instead of implementing a new rule, have you considered using https://palantir.github.io/tslint/rules/ban/? Also supports custom error messages and looks sufficient

devversion reviewed (commented):
LGTM overall, but not sure about the lint rule

alxhub commented on packages/core/src/util/performance.ts:
```diff
@@ -0,0 +1,17 @@
+    markName: string,
+    markOptions?: PerformanceMarkOptions|undefined,
+    ): PerformanceMark|undefined {
+  return performance && typeof performance.mark === 'function' ?
```
```suggestion
  performance?.mark?.(...);
```

alxhub reviewed (approved):
Reviewed-for: global-approvers

alan-agius4 committed:
build: configure tslint to ban `performance.mark` usages

While `performance.mark` is available on all supported browsers and node.js version this API is not available in JSDOM which is used by Jest and Cloudflare worker.

alan-agius4 committed:
fixup! fix(core): guard usages of `performance.mark`

atscott commented:
This PR was merged into the repository by commit d4bd3f198b9680e8b41f0a957d95a651d6ac4a81.

AndrewKushnir commented on packages/core/src/util/performance.ts:
```diff
@@ -0,0 +1,22 @@
+    markName: string,
+    markOptions?: PerformanceMarkOptions|undefined,
+    ): PerformanceMark|undefined {
+  return performance?.mark?.(markName, markOptions);
```
Sorry for the late comment. I'm wondering if we should make it even safer by checking if the `performance` is defined as a global first, i.e.:

```
return typeof performance !== 'undefined' ? performance.mark?.(markName, markOptions) : undefined;
```

@alan-agius4 WDYT?

alan-agius4 replied:
So far I didn’t encountered any occurrences that requires such check.
